R/parallelization.R

```diff
@@ -0,0 +1,144 @@
+#' parallel computations) on your local machine or across the network.
+#'
+#' \pkg{purrr} requires these to be set up prior to performing any parallel map
+#' operations. It is usual to set daemons once per session. You can leave them
+#' running as they consume little in resources whilst waiting to receive tasks.
+#'
+#' If they are not set, you will see a warning from the \pkg{mirai} package that
+#' it is launching one local daemon - this ensures that the function proceeds,
+#' but offers no actual parallelization and is hence probably not what you would
+#' want.
+#'
+#' `daemons()`arguments:
+#'
+#' * `n`: the number of daemons to launch on your local machine, e.g.
+#'   `daemons(7)`. As a rule of thumb, for maximum efficiency this should be (at
+#'   most) one less than the number of cores on your machine, leaving one core
+#'   for the main R process.
+#' * `url` and `remote`: used to set up and launch daemons for distributed
+#'   computing over the network. See [daemons][mirai::daemons] function
+#'   documentation for more details.
+#' * None: calling `daemons()` with no arguments returns a summary of the
+#'   current connection status and mirai tasks queued and completed.
+#'
+#' For more details, including on further arguments, see
+#' [daemons][mirai::daemons].
+#'
+#' Resetting daemons:
+#'
+#' `daemons(0)` resets and terminates any existing daemons.
+#'
+#' All daemons automatically terminate when you end your session and the
+#' connection drops. For this reason, it is not required to specifically
+#' terminate daemons in this instance, although it is still good practice to do
+#' so.
+#'
+#' ## with() method
+#'
+#' The `with()` method for daemons provides a convenient way of evaluating a
+#' statement with daemons set up for the duration of the statement.
+#'
+#' As an example, the below is evaluated in parallel using 7 daemons:
+#'
+#' \preformatted{
+#' with(daemons(7), {
+#'   1:10 |>
+#'   map(rnorm, n = 10, .parallel = TRUE) |>
+#'   map_dbl(mean, .parallel = TRUE)
+#' })
+#' }
+#'
+#' ## Further documentation
+#'
+#' \pkg{purrr}'s parallelization is powered by \CRANpkg{mirai}, so see the
+#' [mirai introduction and reference](https://shikokuchuo.net/mirai/articles/mirai.html)
+#' for more details.
+#'
+#' @name parallelization
+NULL
+
+mmap_ <- function(.x, .f, .args, .parallel, .progress, .type, error_call) {
+
+  m <- if (isTRUE(.parallel)) {
+    mirai::mirai_map(.x, .f, .args = .args)
+  } else if (is.list(.parallel)) {
+    mirai::mirai_map(.x, .f, as.environment(.parallel), .args = .args)
+  } else {
+    cli::cli_abort(
+      "'.parallel' must be TRUE/FALSE or a list, not a {.obj_type_friendly {(.parallel)}}.",
+      call = error_call
+    )
+  }
+
+  options <- c(".stop", if (isTRUE(.progress)) ".progress")
+  x <- withCallingHandlers(
+    mirai::collect_mirai(m, options = options),
+    error = function(cnd) {
+      location <- cnd$location
```

Comment:
Maybe extract out `with_parallel_indexed_errors()` to be put near `with_indexed_errors()` so it is easier for us to keep them in sync in the future?

Reply:
I had a go, but it seems cleaner to keep as is - this is as the interrupt handler requires a reference to the mirai object to be passed in, which would then be inconsistent with having a general `expr` in the fashion of `with_indexed_errors()`.

Reply:
Actually I've taken your suggestion in [fda5abb](https://github.com/tidyverse/purrr/pull/1163/commits/fda5abb85901d18be98c8d715b42dd5afe3e70fa). I pass in an `interrupt_expr` as well as the main `expr`.
It does help to separate out the main logic. Thanks.